### DISCUSSION TITLE
Missing docs on createResetPasswordRequest and resetPassword

### DISCUSSION THREAD

**Discussion by iBobik** [contember/admin#652] [Support]
How to create reset password form?

Missing docs what means mutation options, what token to use and expected workflow?

Also default mail template should be more useful - e.g. do not ask for token copy, but better to click on link like {{ baseUrl }}/login?resetToken={{ token }}.

**Comment by jonasnobile** [comment]
## Step 1: Create a Password Reset Request Form

### `createResetPasswordRequest` Mutation:
- **Email Requirement**: The email entered must exist in the tenant's database.
- **Optional Fields**: You can specify additional options like `mailProject` or `mailVariant` for customizing the email template.
- **Example Mutation**:
  ```gql
  mutation {
    createResetPasswordRequest(
      email: "user@example.com",
      options: {mailProject: "project-name", mailVariant: "variant-1"}
    ) {
      __typename
      error {
        code
        developerMessage
      }
      ok
    }
  }
  ```

## Step 2: User Receives a Password Reset Email

- **Process**: After submitting the form, the user will receive an email with a token to reset their password.
- **Customizing Email Template**: Modify the email template to suit your user's needs.

### Adding a Mail Template Mutation:
- **Customization Options**: Specify the `projectSlug`, `subject`, `type`, and `variant` for the template.
- **Example Mutation**:
  ```gql
  mutation {
    addMailTemplate(
      template: {
        content: "<p>Hello,</p><p>A password reset was requested for your account ({{email}}):<br><a href=\"https://your-domain.com/reset-password/?token={{token}}\">Set new password</p>",
        projectSlug: "project-slug",
        subject: "Reset Password",
        type: RESET_PASSWORD_REQUEST,
        useLayout: true,
        variant: "en"
      }
    ) {
      error {
        code
        developerMessage
      }
      ok
    }
  }
  ```

## Additional Mutations

### Resetting the Password:
- **Usage**: After receiving the token, users can set a new password.
- **Example Mutation**:
  ```gql
  mutation {
    resetPassword(password: "new-password", token: "received-token") {
      error {
        code
        developerMessage
      }
      ok
    }
  }
  ```

### Creating a Login Token:
- **Example Mutation**:
  ```gql
  mutation {
    createGlobalApiKey(description: "Login token", roles: ["login"]) {
      ok
      error {
        code
        developerMessage
      }
      result {
        apiKey {
          token
        }
      }
    }
  }
  ```

**Reply by iBobik** [reply]
Thank you.

Could you clarify in docs what type of token and permission each mutation requires?

createGlobalApiKey requires admin permission? Is it useable for passwordless login? Users asks to login - enters email, sends to the custom backend, it calls createGlobalApiKey and sends email with login link with this token.

But better would be to have mutation in Contember for this.

**Reply by jonasnobile** [reply]
We definitely want to add more information about tokens to our documentation. In the meantime, here is what you need to implement passwordless login.

If you want to implement passwordless login, you want to use `createSessionToken` mutation. 

```gql
mutation {
  createSessionToken(email: "user@example.com") {
    error {
      code
      developerMessage
    }
    ok
    result {
      token
    }
  }
}
```

You will need a `project admin` token for both mutations (`createGlobalApiKey`, `createSessionToken`). Getting the `project admin` token in Contember Cloud is a bit tricky. The easiest way to generate the token is via project admin session token.

1. Get the session token of the admin who created the project in the cloud (sign in to admin and get it from cookie `CONTEMBER_TOKEN`).
2. Create a new global API key with this token
```gql
mutation CreateGlobalApiKey {
  createGlobalApiKey(description: "Project admin", roles: ["project_admin"]) {
    ok
    error {
      code
      developerMessage
    }
    result {
      apiKey {
        token
      }
    }
  }
}
```
3. You can use the Project Admin token on your endpoint to generate the session token.


Is `createSessionToken` mutation the one you were looking for?
